### ISSUE TITLE
Connect() randomly blocks

### ISSUE THREAD

**Issue by vladimirvisage** [theophilusx/ssh2-sftp-client#287]
I use ssh2-sftp-client to get files from an sftp. Sometimes the process blocks, a promise is not returned back. With few logs I found out it was `connect()`, below are the debugs logs (when it fails and when it works) and the code used. There is no timeout, which is stranger.

Do you have any clue why it fails but most importantly why it blocks without timeout? Default is 20000.

ssh2-sftp-client version: 5.1.2
node version: 12.18.4

```js
  logger.debug('New SFTP instance created, connecting...');
  await sftpClient.connect({
    host,
    username,
    privateKey,
    password,
    debug: (msg) => {
      if (msg.startsWith('CLIENT')) {
        logger.debug('debug sftp', { msg });
      } else {
        logger.debug('debug sftp verbose', { msg });
      }
    },
  });
  logger.debug('connected to SFTP');
``` 

The case when the connect method blocks (does not work):
```
@timestamp,message,msg
2020-12-08 02:41:02.372,debug sftp verbose,DEBUG: Client: Connected
2020-12-08 02:41:02.221,debug sftp verbose,DEBUG: Outgoing: Writing CHANNEL_CLOSE (0)
2020-12-08 02:41:02.220,debug sftp verbose,DEBUG: Client: Trying sftp2.domainReplacedHere.net on port 22 ...
2020-12-08 02:41:02.219,Opening SFTP connection
```

The case when it works (I have my 2 logs + debugs logs from ssh2):
```
2020-12-08 02:40:04.315,connected to SFTP,
2020-12-08 02:40:04.314,debug sftp verbose,DEBUG: Parser: IN_PACKET
2020-12-08 02:40:04.314,debug sftp verbose,DEBUG: Parser: Decrypting
2020-12-08 02:40:04.159,debug sftp verbose,DEBUG: Parser: Decrypting
2020-12-08 02:40:04.159,debug sftp verbose,DEBUG: Parser: HMAC size:32
2020-12-08 02:40:04.159,debug sftp verbose,DEBUG: Parser: IN_PACKETDATAVERIFY
2020-12-08 02:40:04.159,debug sftp verbose,DEBUG: Parser: Verifying MAC
2020-12-08 02:40:04.159,debug sftp verbose,DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
2020-12-08 02:40:04.159,debug sftp verbose,"DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (0)"
2020-12-08 02:40:04.159,debug sftp,CLIENT[sftp]: SFTP connection established
2020-12-08 02:40:04.159,debug sftp,CLIENT[sftp]: Connected - do stage 2 setup
2020-12-08 02:40:04.159,debug sftp verbose,DEBUG[SFTP]: Outgoing: Writing REALPATH
2020-12-08 02:40:04.159,debug sftp verbose,DEBUG: Outgoing: Writing CHANNEL_DATA (0)
2020-12-08 02:40:04.159,debug sftp verbose,DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
2020-12-08 02:40:04.158,debug sftp verbose,DEBUG: Parser: IN_PACKET
2020-12-08 02:40:04.158,debug sftp verbose,DEBUG: Parser: Decrypting
2020-12-08 02:40:04.158,debug sftp verbose,"DEBUG: Parser: pktLen:172,padLen:8,remainLen:160"
2020-12-08 02:40:04.158,debug sftp verbose,DEBUG: Parser: IN_PACKETDATA
2020-12-08 02:40:04.002,debug sftp verbose,DEBUG: Parser: HMAC size:32
2020-12-08 02:40:04.002,debug sftp verbose,DEBUG: Parser: IN_PACKETDATAVERIFY
2020-12-08 02:40:04.002,debug sftp verbose,DEBUG: Parser: Verifying MAC
2020-12-08 02:40:04.002,debug sftp verbose,DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
2020-12-08 02:40:04.002,debug sftp verbose,"DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_SUCCESS (0)"
2020-12-08 02:40:04.002,debug sftp verbose,DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
2020-12-08 02:40:04.002,debug sftp verbose,DEBUG: Outgoing: Writing CHANNEL_DATA (0)
2020-12-08 02:40:04.001,debug sftp verbose,DEBUG: Parser: IN_PACKET
2020-12-08 02:40:04.001,debug sftp verbose,DEBUG: Parser: Decrypting
2020-12-08 02:40:04.001,debug sftp verbose,"DEBUG: Parser: pktLen:28,padLen:18,remainLen:16"
2020-12-08 02:40:04.001,debug sftp verbose,DEBUG: Parser: IN_PACKETDATA
2020-12-08 02:40:04.001,debug sftp verbose,DEBUG: Parser: Decrypting
2020-12-08 02:40:04.001,debug sftp verbose,DEBUG: Parser: HMAC size:32
2020-12-08 02:40:04.001,debug sftp verbose,DEBUG: Parser: IN_PACKETDATAVERIFY
2020-12-08 02:40:04.001,debug sftp verbose,DEBUG: Parser: Verifying MAC
2020-12-08 02:40:04.001,debug sftp verbose,DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
2020-12-08 02:40:04.001,debug sftp verbose,"DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_WINDOW_ADJUST (0, 2097152)"
2020-12-08 02:40:04.001,debug sftp verbose,DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
2020-12-08 02:40:04.001,debug sftp verbose,DEBUG: Parser: IN_PACKET
2020-12-08 02:40:04.001,debug sftp verbose,DEBUG: Parser: Decrypting
2020-12-08 02:40:04.001,debug sftp verbose,"DEBUG: Parser: pktLen:12,padLen:6,remainLen:0"
2020-12-08 02:40:04.001,debug sftp verbose,DEBUG: Parser: IN_PACKETDATA
2020-12-08 02:40:03.845,debug sftp verbose,DEBUG: Parser: IN_PACKET
2020-12-08 02:40:03.845,debug sftp verbose,DEBUG: Parser: Decrypting
2020-12-08 02:40:03.845,debug sftp verbose,"DEBUG: Parser: pktLen:28,padLen:10,remainLen:16"
2020-12-08 02:40:03.845,debug sftp verbose,DEBUG: Parser: IN_PACKETDATA
2020-12-08 02:40:03.845,debug sftp verbose,DEBUG: Parser: Decrypting
2020-12-08 02:40:03.845,debug sftp verbose,DEBUG: Parser: HMAC size:32
2020-12-08 02:40:03.845,debug sftp verbose,DEBUG: Parser: IN_PACKETDATAVERIFY
2020-12-08 02:40:03.845,debug sftp verbose,DEBUG: Parser: Verifying MAC
2020-12-08 02:40:03.845,debug sftp verbose,DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
2020-12-08 02:40:03.845,debug sftp verbose,"DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_OPEN_CONFIRMATION"
2020-12-08 02:40:03.845,debug sftp verbose,"DEBUG: Outgoing: Writing CHANNEL_REQUEST (0, subsystem: sftp)"
2020-12-08 02:40:03.845,debug sftp verbose,DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
2020-12-08 02:40:03.653,debug sftp verbose,DEBUG: Parser: IN_PACKET
2020-12-08 02:40:03.653,debug sftp verbose,DEBUG: Parser: Decrypting
2020-12-08 02:40:03.653,debug sftp verbose,"DEBUG: Parser: pktLen:924,padLen:11,remainLen:912"
2020-12-08 02:40:03.653,debug sftp verbose,DEBUG: Parser: IN_PACKETDATA
2020-12-08 02:40:03.653,debug sftp verbose,DEBUG: Parser: Decrypting
2020-12-08 02:40:03.653,debug sftp verbose,DEBUG: Parser: HMAC size:32
2020-12-08 02:40:03.653,debug sftp verbose,DEBUG: Parser: IN_PACKETDATAVERIFY
2020-12-08 02:40:03.653,debug sftp verbose,DEBUG: Parser: Verifying MAC
2020-12-08 02:40:03.653,debug sftp verbose,DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
2020-12-08 02:40:03.653,debug sftp verbose,"DEBUG: Parser: IN_PACKETDATAAFTER, packet: GLOBAL_REQUEST (hostkeys-00@openssh.com)"
2020-12-08 02:40:03.653,debug sftp verbose,DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
2020-12-08 02:40:03.391,debug sftp verbose,DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
2020-12-08 02:40:03.391,debug sftp verbose,"DEBUG: Parser: IN_PACKETDATAAFTER, packet: USERAUTH_SUCCESS"
2020-12-08 02:40:03.391,debug sftp verbose,"DEBUG: Outgoing: Writing CHANNEL_OPEN (0, session)"
2020-12-08 02:40:03.391,debug sftp verbose,DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
2020-12-08 02:40:03.390,debug sftp verbose,DEBUG: Parser: IN_PACKET
2020-12-08 02:40:03.390,debug sftp verbose,DEBUG: Parser: Decrypting
2020-12-08 02:40:03.390,debug sftp verbose,"DEBUG: Parser: pktLen:12,padLen:10,remainLen:0"
2020-12-08 02:40:03.390,debug sftp verbose,DEBUG: Parser: IN_PACKETDATA
2020-12-08 02:40:03.390,debug sftp verbose,DEBUG: Parser: HMAC size:32
2020-12-08 02:40:03.390,debug sftp verbose,DEBUG: Parser: IN_PACKETDATAVERIFY
2020-12-08 02:40:03.390,debug sftp verbose,DEBUG: Parser: Verifying MAC
2020-12-08 02:40:03.206,debug sftp verbose,DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
2020-12-08 02:40:03.206,debug sftp verbose,"DEBUG: Parser: IN_PACKETDATAAFTER, packet: USERAUTH_FAILURE"
2020-12-08 02:40:03.206,debug sftp verbose,DEBUG: Client: none auth failed
2020-12-08 02:40:03.206,debug sftp verbose,DEBUG: Outgoing: Writing USERAUTH_REQUEST (password)
2020-12-08 02:40:03.206,debug sftp verbose,DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
2020-12-08 02:40:03.205,debug sftp verbose,DEBUG: Parser: IN_PACKET
2020-12-08 02:40:03.205,debug sftp verbose,DEBUG: Parser: Decrypting
2020-12-08 02:40:03.205,debug sftp verbose,"DEBUG: Parser: pktLen:44,padLen:19,remainLen:32"
2020-12-08 02:40:03.205,debug sftp verbose,DEBUG: Parser: IN_PACKETDATA
2020-12-08 02:40:03.205,debug sftp verbose,DEBUG: Parser: Decrypting
2020-12-08 02:40:03.205,debug sftp verbose,DEBUG: Parser: HMAC size:32
2020-12-08 02:40:03.205,debug sftp verbose,DEBUG: Parser: IN_PACKETDATAVERIFY
2020-12-08 02:40:03.205,debug sftp verbose,DEBUG: Parser: Verifying MAC
2020-12-08 02:40:03.048,debug sftp verbose,DEBUG: Parser: IN_PACKET
2020-12-08 02:40:03.048,debug sftp verbose,DEBUG: Parser: Decrypting
2020-12-08 02:40:03.048,debug sftp verbose,"DEBUG: Parser: pktLen:28,padLen:10,remainLen:16"
2020-12-08 02:40:03.048,debug sftp verbose,DEBUG: Parser: IN_PACKETDATA
2020-12-08 02:40:03.048,debug sftp verbose,DEBUG: Parser: Decrypting
2020-12-08 02:40:03.048,debug sftp verbose,DEBUG: Parser: HMAC size:32
2020-12-08 02:40:03.048,debug sftp verbose,DEBUG: Parser: IN_PACKETDATAVERIFY
2020-12-08 02:40:03.048,debug sftp verbose,DEBUG: Parser: Verifying MAC
2020-12-08 02:40:03.048,debug sftp verbose,DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
2020-12-08 02:40:03.048,debug sftp verbose,"DEBUG: Parser: IN_PACKETDATAAFTER, packet: SERVICE_ACCEPT"
2020-12-08 02:40:03.048,debug sftp verbose,DEBUG: Outgoing: Writing USERAUTH_REQUEST (none)
2020-12-08 02:40:03.048,debug sftp verbose,DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
2020-12-08 02:40:02.893,debug sftp verbose,DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
2020-12-08 02:40:02.892,debug sftp verbose,DEBUG: Verifying signature
2020-12-08 02:40:02.892,debug sftp verbose,DEBUG: Outgoing: Writing NEWKEYS
2020-12-08 02:40:02.892,debug sftp verbose,DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
2020-12-08 02:40:02.892,debug sftp verbose,DEBUG: Parser: IN_PACKET
2020-12-08 02:40:02.892,debug sftp verbose,"DEBUG: Parser: pktLen:12,padLen:10,remainLen:8"
2020-12-08 02:40:02.892,debug sftp verbose,DEBUG: Parser: IN_PACKETDATA
2020-12-08 02:40:02.892,debug sftp verbose,"DEBUG: Parser: IN_PACKETDATAAFTER, packet: NEWKEYS"
2020-12-08 02:40:02.892,debug sftp verbose,DEBUG: Outgoing: Writing SERVICE_REQUEST (ssh-userauth)
2020-12-08 02:40:02.891,debug sftp verbose,DEBUG: Parser: IN_PACKET
2020-12-08 02:40:02.891,debug sftp verbose,"DEBUG: Parser: pktLen:188,padLen:8,remainLen:184"
2020-12-08 02:40:02.891,debug sftp verbose,DEBUG: Parser: IN_PACKETDATA
2020-12-08 02:40:02.891,debug sftp verbose,"DEBUG: Parser: IN_PACKETDATAAFTER, packet: KEXECDH_REPLY"
2020-12-08 02:40:02.891,debug sftp verbose,DEBUG: Checking host key format
2020-12-08 02:40:02.891,debug sftp verbose,DEBUG: Checking signature format
2020-12-08 02:40:02.891,debug sftp verbose,DEBUG: Verifying host fingerprint
2020-12-08 02:40:02.891,debug sftp verbose,DEBUG: Host accepted by default (no verification)
2020-12-08 02:40:02.724,debug sftp verbose,"DEBUG: (local) Client->Server ciphers: aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm,aes128-gcm@openssh.com,aes256-gcm,aes256-gcm@openssh.com"
2020-12-08 02:40:02.724,debug sftp verbose,"DEBUG: (remote) Client->Server ciphers: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com"
2020-12-08 02:40:02.724,debug sftp verbose,DEBUG: Client->Server Cipher: aes128-ctr
2020-12-08 02:40:02.724,debug sftp verbose,"DEBUG: (local) Server->Client ciphers: aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm,aes128-gcm@openssh.com,aes256-gcm,aes256-gcm@openssh.com"
2020-12-08 02:40:02.724,debug sftp verbose,"DEBUG: (remote) Server->Client ciphers: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com"
2020-12-08 02:40:02.724,debug sftp verbose,DEBUG: Server->Client Cipher: aes128-ctr
2020-12-08 02:40:02.724,debug sftp verbose,"DEBUG: (local) Client->Server HMAC algorithms: hmac-sha2-256,hmac-sha2-512,hmac-sha1"
2020-12-08 02:40:02.724,debug sftp verbose,"DEBUG: (remote) Client->Server HMAC algorithms: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1"
2020-12-08 02:40:02.724,debug sftp verbose,DEBUG: Client->Server HMAC algorithm: hmac-sha2-256
2020-12-08 02:40:02.724,debug sftp verbose,"DEBUG: (local) Server->Client HMAC algorithms: hmac-sha2-256,hmac-sha2-512,hmac-sha1"
2020-12-08 02:40:02.724,debug sftp verbose,"DEBUG: (remote) Server->Client HMAC algorithms: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1"
2020-12-08 02:40:02.724,debug sftp verbose,DEBUG: Server->Client HMAC algorithm: hmac-sha2-256
2020-12-08 02:40:02.724,debug sftp verbose,"DEBUG: (local) Client->Server compression algorithms: none,zlib@openssh.com,zlib"
2020-12-08 02:40:02.724,debug sftp verbose,"DEBUG: (remote) Client->Server compression algorithms: none,zlib@openssh.com"
2020-12-08 02:40:02.724,debug sftp verbose,DEBUG: Client->Server compression algorithm: none
2020-12-08 02:40:02.724,debug sftp verbose,"DEBUG: (local) Server->Client compression algorithms: none,zlib@openssh.com,zlib"
2020-12-08 02:40:02.724,debug sftp verbose,"DEBUG: (remote) Server->Client compression algorithms: none,zlib@openssh.com"
2020-12-08 02:40:02.724,debug sftp verbose,DEBUG: Server->Client compression algorithm: none
2020-12-08 02:40:02.724,debug sftp verbose,DEBUG: Outgoing: Writing KEXECDH_INIT
2020-12-08 02:40:02.724,debug sftp verbose,DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
2020-12-08 02:40:02.723,debug sftp verbose,DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
2020-12-08 02:40:02.723,debug sftp verbose,DEBUG: Parser: IN_PACKET
2020-12-08 02:40:02.723,debug sftp verbose,"DEBUG: Parser: pktLen:972,padLen:10,remainLen:968"
2020-12-08 02:40:02.723,debug sftp verbose,DEBUG: Parser: IN_PACKETDATA
2020-12-08 02:40:02.723,debug sftp verbose,"DEBUG: Parser: IN_PACKETDATAAFTER, packet: KEXINIT"
2020-12-08 02:40:02.723,debug sftp verbose,DEBUG: Comparing KEXINITs ...
2020-12-08 02:40:02.723,debug sftp verbose,"DEBUG: (local) KEX algorithms: curve25519-sha256@libssh.org,curve25519-sha256,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group14-sha1"
2020-12-08 02:40:02.723,debug sftp verbose,"DEBUG: (remote) KEX algorithms: curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha1"
2020-12-08 02:40:02.723,debug sftp verbose,DEBUG: KEX algorithm: curve25519-sha256@libssh.org
2020-12-08 02:40:02.723,debug sftp verbose,"DEBUG: (local) Host key formats: ssh-ed25519,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,ssh-rsa"
2020-12-08 02:40:02.723,debug sftp verbose,"DEBUG: (remote) Host key formats: ssh-rsa,rsa-sha2-512,rsa-sha2-256,ecdsa-sha2-nistp256,ssh-ed25519"
2020-12-08 02:40:02.723,debug sftp verbose,DEBUG: Host key format: ssh-ed25519
2020-12-08 02:40:02.569,debug sftp verbose,DEBUG: Outgoing: Writing KEXINIT
2020-12-08 02:40:02.568,debug sftp verbose,DEBUG: Parser: IN_INIT
2020-12-08 02:40:02.568,debug sftp verbose,DEBUG: Parser: IN_GREETING
2020-12-08 02:40:02.568,debug sftp verbose,DEBUG: Parser: IN_HEADER
2020-12-08 02:40:02.568,debug sftp verbose,DEBUG: Remote ident: 'SSH-2.0-OpenSSH_7.2p2 Ubuntu-4ubuntu2.10'
2020-12-08 02:40:02.406,debug sftp verbose,DEBUG: Client: Connected
2020-12-08 02:40:02.249,debug sftp verbose,DEBUG: Local ident: 'SSH-2.0-ssh2js0.4.10'
2020-12-08 02:40:02.249,debug sftp verbose,DEBUG: Client: Trying sftp2.domainReplacedHere.net on port 22 ...
2020-12-08 02:40:02.249,debug sftp verbose,DEBUG: Outgoing: Writing CHANNEL_CLOSE (0)
2020-12-08 02:40:02.248,Opening SFTP connection,
```

**Comment by vladimirvisage** [comment]
Seems related to this issue: https://github.com/mscdex/ssh2/issues/928

So `connect()` would benefit from listening to events `close` and `end`.

Would you confirm this analysis?

**Comment by theophilusx** [comment]
I'm not sure. The connect promise does have a listener for an error
event, but not a close event. What you can do is add a close listener
and see if a close event is being raised when your connect hangs. If you
can confirm there is actually a close event being raised during the
connect, it would be easy to add a reject if it occurs during connect.
Strictly speaking, I think ssh2 should probably raise an error event if
the connection is suddenly closed by the remote server. However, as the
maintainer is doing a complete re-write of that module and deprecating
the old ssh2-streams module, I doubt we will se any change.

All my spare time at present is devoted to trying to re-write
ssh2-sftp-client to work with the new ssh2 module as I'm hoping this
will also fix the node v14 issues with the ssh2 module.

You might also want to try updating to the current version of
ssh2-sftp-client. There were some changes made to listeners, but I
cannot remember if that was before or after 5.2.1 (current version is
5.3.2).

vladimirvisage <notifications@github.com> writes:

> Seems related to this issue: https://github.com/mscdex/ssh2/issues/928
>
> So `connect()` would benefit from listening to events `close` and `end`.
>
> Would you confirm this analysis?


--
Tim Cross

**Comment by vladimirvisage** [comment]
Thank you a lot for your vision on it.

The 5.3.2 has a little change on listeners but still not listening to "close"
event.

From my test (listening to close & end events), the issue is detected
when it happens. So I have done a wrapper to the connect method.
I will check if this works and take it as a fix for now until a new version
based on the new ssh2 module is made. Do you have a clue about the 
release date?

Until then, if this fix works a good thing would be to fix in 5.3.3 this event.
If you are busy with the v6 I can submit a proposal.

Here is the workaround as a wrapper (not the lib fix):

```
function connectSftp(sftpClient, connectOptions) {
  const connectionPromise = sftpClient.connect(connectOptions).catch((err) => {
    logger.debug('SSH failed', { err });
  });

  const safeConnectionPromise = new Promise((resolve, reject) => {
    const ssh = sftpClient.client;

    const onEarlyEnd = () => {
      logger.debug('SSH early close / end');
      removeListeners();
      reject(new Error('SSH early close / end'));
    };
    const onReady = () => {
      logger.debug('SSH ready');
      removeListeners();
      resolve();
    };
    const removeListeners = () => {
      ssh.off('close', onEarlyEnd);
      ssh.off('end', onEarlyEnd);
      ssh.off('ready', onReady);
      ssh.off('error', reject);
    };

    ssh.on('ready', onReady);
    ssh.on('error', reject);
    ssh.on('close', onEarlyEnd);
    ssh.on('end', onEarlyEnd);
  });

  // Double check connection state, connect() can miss a early close of the connection giving infinite promise.
  return Promise.all([connectionPromise, safeConnectionPromise]);
}
```

**Comment by theophilusx** [comment]
I don't know when a new version based on the new version of ssh2 will be
released and as yet, don't know if the new ssh2 will change the
behaviour of ssh2 wrt close v error events (or even if it can - this
could easily be related to the inherent behaviour of the remote sftp
server). A new version of ssh2-sftp-client based on the new ssh2 won't
be released before ssh2 is released, but I do hope to be in a position
to do so quickly when it is.

I'm more than happy to consider any pull requests submitted. The code in
the connect method is my least favourite bit of code in
ssh2-sftp-client. The 'retry' implementation is badly in need of
re-factoring (and it has a couple of minor bugs). You also need to be
very careful about removing listeners added during the connect process
(but not removing the global listeners which need to stay in place). It
is very easy to trip node's possible memory leak warning  if you fail to do
this when you reuse the same ssh2-sftp-client object for multiple
connections. Mixing Promises with an event model tends to cause a bit of
'tension' and there are a few edge cases that are hard to
detect/reproduce. There appears to be some variability in how servers
deal with events that complicate matters. It will probably also be
necessary to extent the tests with additional tests (I've been meaning
to add additional tests to ensure we don't see a slow build up of event
listeners etc).

BTW it would also be great to find out a bit more about what sftp server
is at the far end. The issue you have encountered is not common and I've
not been able to find a test server where I can reproduce the behaviour
(i.e. issue a close/end event without an error event when connection
terminated unexpectedly by remote server). Would be good to document
unusual/unexpected behaviour of some sftp servers.


vladimirvisage <notifications@github.com> writes:

> Thank you a lot for your vision on it.
>
> The 5.3.2 has a little change on listeners but still not listening to "close"
> event.
>
> From my test (listening to close & end events), the issue is detected
> when it happens. So I have done a wrapper to the connect method.
> I will check if this works and take it as a fix for now until a new version
> based on the new ssh2 module is made. Do you have a clue about the
> release date?
>
> Until then, if this fix works a good thing would be to fix in 5.3.3 this event.
> If you are busy with the v6 I can submit a proposal.
>
> Here is the workaround as a wrapper (not the lib fix):
>
> ```
> function connectSftp(sftpClient, connectOptions) {
>   const connectionPromise = sftpClient.connect(connectOptions).catch((err) => {
>     logger.debug('SSH failed', { err });
>   });
>
>   const safeConnectionPromise = new Promise((resolve, reject) => {
>     const ssh = sftpClient.client;
>
>     const onEarlyEnd = () => {
>       logger.debug('SSH early close / end');
>       removeListeners();
>       reject(new Error('SSH early close / end'));
>     };
>     const onReady = () => {
>       logger.debug('SSH ready');
>       removeListeners();
>       resolve();
>     };
>     const removeListeners = () => {
>       ssh.off('close', onEarlyEnd);
>       ssh.off('end', onEarlyEnd);
>       ssh.off('ready', onReady);
>       ssh.off('error', reject);
>     };
>
>     ssh.on('ready', onReady);
>     ssh.on('error', reject);
>     ssh.on('close', onEarlyEnd);
>     ssh.on('end', onEarlyEnd);
>   });
>
>   // Double check connection state, connect() can miss a early close of the connection giving infinite promise.
>   return Promise.all([connectionPromise, safeConnectionPromise]);
> }
> ```


--
Tim Cross

**Comment by theophilusx** [comment]
I have made some changes to add both close and end listeners in a more consistent manner in the version-6 branch. Currently, the retry feature is missing as I'm still refactoring that bit of the code. However, you could try it out to see if that fixes your problem. NOTE that you will need to change the ssh2 version to use "^0.8.9" and not use the current ssh2 github master branch as it definitely has problems with fastPut (and probably fastGet and other functions). The ssh2 master branch is definitely not yet ready. 

As soon as I've gathered some data to submit an update to ssh2 regarding fastPut() problems, I will return to refactoring the connect function to add back the retry feature. If it is then all stable enough, I may use that branch to release an updated 5.x version as it looks like the new ssh2 version might be some time away.

**Comment by theophilusx** [comment]
The code iin the version-6 branch has significant improvements in handling close, end and error events and I have now added back support for retrying connections. I think it is better than the 5.3.x code, so please give it a try and see if it fixes the issues you were having.

**Comment by theophilusx** [comment]
Have released v6.0.0, which I hope has fixed the issue of scripts hanging due to the remote server abruptly closing the connection without issuing an error event. Therefore, closing this issue.
